### PR TITLE
Chore: Fully deprecate `objc2::rc::Id`

### LINKED ISSUE #617: Rename `objc2::rc::Id` to `objc2::rc::Retained`
`Id<T>` is a leftover from `objc_id`, but that name doesn't actually convey any meaning to an outsider, and has only weak relevance to the concept in Objective-C (it somewhat matches `id`, but that's actually closer to `AnyObject`). Additionally, the shorthand "ID" is usually used to refer to an "Identity Document" or an "Identifier", both of which this most definitely is not.

Instead, we should use the name `Retained<T>`, which more clearly communicates that it is a retained smart pointer. (The concept of `retain`/`release` is one that I'd more expect previous and future Objective-C developers to understand).

Alternative names:
- `Arc<T>`, would match well with the ARC i.e. Automatic Reference Counting concept in Objective-C, but conflicts with `std::sync::Arc`. Still, it's what [`fruity`](https://github.com/nvzqz/fruity) does.
- `R<T>` as a shorthand. Also what [`cidre`](https://github.com/yury/cidre) does.
- `Strong<T>`.
- `Object<T>`.
- `ArcObject<T>`
- ?

Migration plan:
- `objc2 v0.5.2`: Introduce `Retained<T>` as an alias of `Id<T>`. EDIT: Done in b8adc2b36e9720a283cecef5209d14be818052d6.
- Continuously: Convert internals to use `Retained<T>` instead of `Id<T>`. EDIT: Done in d097cd26fab389d66ff766b1f719a32426962ab2.
- Figure out what to do about `#[method_id(...)]`: https://github.com/madsmtm/objc2/issues/457
- `objc2 v0.6.0`: Deprecate the name `Id<T>`.
- ?: Remove `Id<T>` (maybe never, the cost of keeping the alias is basically zero).

This also affects:
- `WeakId<T>`, should be renamed to `RetainedWeak<T>`? Or just `objc2::rc::Weak<T>`, just like we have both `std::sync::Weak<T>` and `std::rc::Weak<T>`?
- `msg_send_id!` / `#[method_id(...))]`, I'm going to keep these names as `id` is still a concept in Objective-C that refers to "an object", and calling them `_retain` would be wrong (as whether or not to retain is). Maybe `_object` or `_` would be better, but it's not as glaring a problem.
- `IsIdCloneable`, should be renamed to `IsRetainedCloneable`? Or maybe don't do anything here, and instead wait for https://github.com/madsmtm/objc2/issues/563.
- `DefaultId`, `IdFromIterator` and `IdIntoIterator` should all be renamed.

CC @simlay @silvanshade, I'm making this change far too late in the project's lifetime, but I believe it to be the right thing to do, even considering the maintenance cost to dependencies. Do you have alternative naming suggestions, or other opinions?

### PR DESCRIPTION
**Pull request by PaulDance**
As part of #617.

With this, 0.6 should be in a good position to be delivered, or are there still a lot of things to do? What I see in https://github.com/madsmtm/objc2/milestone/6 seems that it could be optional while the important changes that are already completed would be nice if published and back-porting things I would need in a fork would be cumbersome.

### CODE REVIEW

**Review by madsmtm — CHANGES_REQUESTED** [review]
> As far as I can tell, this should be the last remaining part required in order to close #617.

I think I'd also like to figure out `#[method_id(...)]`, I've added mention of https://github.com/madsmtm/objc2/issues/457 to the issue description.

> With this, 0.6 should be in a good position to be delivered, or are there still a lot of things to do? What I see in https://github.com/madsmtm/objc2/milestone/6 seems that it could be optional while the important changes that are already completed would be nice if published and back-porting things I would need in a fork would be cumbersome.

Ugh, I'm so bad at deciding when to release, in my defence things are complicated by us having a fairly large amount of dependents, and thus infrequent breaking versions are desirable for the ecosystem in general. But you are right, we are _close_, the target date I've set for the milestone (end of the year) _should_ be accurate now, the remaining stuff that I really want to get in is:
- Proper enum prefix stripping, having to write `NSWindowSharingType::NSWindowSharingNone` is really ugly when `NSWindowSharingType::None` could do.
- The cargo features stuff (removing `"all"` in favour of enabling things by default + std features shouldn't enable sub-features), since it's a breaking change.
- The autogenerated thunks for functions, since it's also a breaking change, and it's needed for nicer Metal integration (allows `objc2-metal` to really shine over `metal-rs`).

(So if you want to explore some of those, then feel free to ;) )

File: crates/objc2/src/rc/retained.rs [hunk 1]
```diff
@@ -146,11 +146,8 @@ pub struct Retained<T: ?Sized> {
...
 /// Short type-alias to [`Retained`].
 ///
-/// This type is soft-deprecated, and will be fully deprecated in `v0.6.0`,
-/// see [#617].
-///
-/// [`Retained<T>`]: Retained
-/// [#617]: https://github.com/madsmtm/objc2/issues/617
+/// This is fully deprecated since `v0.6.0`, use [`Retained`] instead.
+#[deprecated(since = "0.6.0", note = "Renamed to `Retained`.")]
 pub type Id<T> = Retained<T>;
```
[the remark this answers is not in the document]
Comment (PaulDance) [reply crates/objc2/src/rc/retained.rs:151]: Didn't notice this one! Done :slightly_smiling_face:

**Review by madsmtm — APPROVED** [review]
Thanks!

**Comment by PaulDance** [comment]
> I think I'd also like to figure out `#[method_id(...)]`, I've added mention of #457 to the issue description.

OK, changed the description accordingly.

> I'm so bad at deciding when to release

Most definitely understandable. Another way to see things though is that breaking a lot at the same time makes transitioning harder, while more frequent but more focused breakage is somewhat more manageable. Tough choices overall.

> the target date I've set for the milestone (end of the year) should be accurate now

That is more than acceptable, especially considering vacations and stuff. It should also align nicely with a development I will take on again.

> So if you want to explore some of those, then feel free to ;)

I just might, but maybe only the first two :wink:

> Thanks!

You're welcome!

**Comment by madsmtm** [comment]
Update on the timeline I gave: I of course didn't make it by the end of the year, but I _have_ progressed on the things I said I wanted to get in, the main things missing is the cargo features stuff (I have a local WIP branch for this, but it's tricky). So we are getting _closer_...

**Comment by PaulDance** [comment]
Yes, I have seen the recent progress and it's looking good indeed! Regarding the timeline, the projects were big and numerous, so it is not exactly surprising the originally-estimated date was passed. However, considering it should only be by a few days in total and that my use of it is only optional during the development that is only starting now really, I think it should be fine :slightly_smiling_face:

**Pull request merged**